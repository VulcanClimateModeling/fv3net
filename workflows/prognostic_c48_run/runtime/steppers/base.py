import logging

import xarray as xr
from runtime.diagnostics.machine_learning import compute_baseline_diagnostics
from runtime.names import TENDENCY_TO_STATE_NAME
from runtime.types import State

logger = logging.getLogger(__name__)

KG_PER_M2_PER_M = 1000.0


class LoggingMixin:

    rank: int

    def _log_debug(self, message: str):
        if self.rank == 0:
            logger.debug(message)

    def _log_info(self, message: str):
        if self.rank == 0:
            logger.info(message)

    def _print(self, message: str):
        if self.rank == 0:
            print(message)


class BaselineStepper:
    net_moistening = "doesn't matter"

    def __call__(self, time, state):
        return {}, {}, {}, None, {}

    def get_diagnostics(self, state, tendency):
        return compute_baseline_diagnostics(state)

    def get_momentum_diagnostics(self, state, tendency):
        return {}

    def apply(self, state, tendency, dt):
        return apply(state, tendency, dt)


def apply(state: State, tendency: State, dt: float) -> State:
    """Given state and tendency prediction, return updated state.
    Returned state only includes variables updated by ML model."""

    with xr.set_options(keep_attrs=True):
        updated = {}
        for name in tendency:
            state_name = TENDENCY_TO_STATE_NAME.get(name, name)
            updated[state_name] = state[state_name] + tendency[name] * dt
    return updated  # type: ignore


def precipitation_sum(
    physics_precip: xr.DataArray, column_dq2: xr.DataArray, dt: float
) -> xr.DataArray:
    """Return sum of physics precipitation and ML-induced precipitation. Output is
    thresholded to enforce positive precipitation.

    Args:
        physics_precip: precipitation from physics parameterizations [m]
        column_dq2: column-integrated moistening from ML [kg/m^2/s]
        dt: physics timestep [s]

    Returns:
        total precipitation [m]"""
    m_per_mm = 1 / 1000
    ml_precip = -column_dq2 * dt * m_per_mm  # type: ignore
    total_precip = physics_precip + ml_precip
    total_precip = total_precip.where(total_precip >= 0, 0)
    total_precip.attrs["units"] = "m"
    return total_precip


def precipitation_rate(
    precipitation_accumulation: xr.DataArray, dt: float
) -> xr.DataArray:
    """Return precipitation rate from a precipitation accumulation and timestep
    
    Args:
        precipitation_accumulation: precipitation accumulation [m]
        dt: timestep over which accumulation occurred [s]

    Returns:
        precipitation rate [kg/m^s/s]"""
    precipitation_rate: xr.DataArray = (
        KG_PER_M2_PER_M * precipitation_accumulation / dt  # type: ignore
    )
    precipitation_rate.attrs["units"] = "kg/m^2/s"
    return precipitation_rate
