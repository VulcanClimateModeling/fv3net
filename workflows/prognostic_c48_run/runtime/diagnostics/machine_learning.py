from typing import MutableMapping, Hashable
import xarray as xr
import logging
from vcm.catalog import catalog
from vcm.cubedsphere import center_and_rotate_xy_winds

logger = logging.getLogger(__name__)

TEMP = "air_temperature"
SPHUM = "specific_humidity"
DELP = "pressure_thickness_of_atmospheric_layer"
PRECIP_RATE = "surface_precipitation_rate"
cp = 1004
gravity = 9.81

State = MutableMapping[Hashable, xr.DataArray]
Diagnostics = MutableMapping[Hashable, xr.DataArray]


def compute_ml_diagnostics(state: State, ml_tendency: State) -> Diagnostics:

    net_moistening = (ml_tendency["dQ2"] * state[DELP] / gravity).sum("z")
    physics_precip = state[PRECIP_RATE]

    return dict(
        air_temperature=state[TEMP],
        specific_humidity=state[SPHUM],
        pressure_thickness_of_atmospheric_layer=state[DELP],
        net_moistening=(net_moistening)
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(description="column integrated ML model moisture tendency"),
        net_heating=(ml_tendency["dQ1"] * state[DELP] / gravity * cp)
        .sum("z")
        .assign_attrs(units="W/m^2")
        .assign_attrs(description="column integrated ML model heating"),
        water_vapor_path=(state[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="mm")
        .assign_attrs(description="column integrated water vapor"),
        physics_precip=(physics_precip)
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(
            description="surface precipitation rate due to parameterized physics"
        ),
    )


def compute_ml_momentum_diagnostics(state: State, tendency: State) -> Diagnostics:
    delp = state[DELP]

    dQu = tendency.get("dQu", xr.zeros_like(delp))
    dQv = tendency.get("dQv", xr.zeros_like(delp))
    column_integrated_dQu = _mass_average(dQu, delp, "z")
    column_integrated_dQv = _mass_average(dQv, delp, "z")
    return dict(
        column_integrated_dQu=column_integrated_dQu.assign_attrs(
            units="m s^-2",
            description="column integrated zonal wind tendency due to ML",
        ),
        column_integrated_dQv=column_integrated_dQv.assign_attrs(
            units="m s^-2",
            description="column integrated meridional wind tendency due to ML",
        ),
    )


def rename_diagnostics(diags: Diagnostics):
    """Postfix ML output names with _diagnostic and create zero-valued outputs in
    their stead. Function operates in place."""
    ml_tendencies = {
        "net_moistening",
        "net_heating",
        "column_integrated_dQu",
        "column_integrated_dQv",
    }
    ml_tendencies_in_diags = ml_tendencies & set(diags)
    for variable in ml_tendencies_in_diags:
        attrs = diags[variable].attrs
        diags[f"{variable}_diagnostic"] = diags[variable].assign_attrs(
            description=attrs["description"] + " (diagnostic only)"
        )
        diags[variable] = xr.zeros_like(diags[variable]).assign_attrs(attrs)


def compute_nudging_diagnostics(
    state: State, nudging_tendency: State, label: str = "_tendency_due_to_nudging"
) -> Diagnostics:
    """
    Compute diagnostic variables for nudging"""

    diags: Diagnostics = {}

    net_moistening = (
        (nudging_tendency[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(description="column integrated moistening due to nudging")
    )
    net_heating = (
        (nudging_tendency[TEMP] * state[DELP] / gravity * cp)
        .sum("z")
        .assign_attrs(units="W/m^2")
        .assign_attrs(description="column integrated heating due to nudging")
    )
    water_vapor_path = (
        (state[SPHUM] * state[DELP] / gravity)
        .sum("z")
        .assign_attrs(units="mm")
        .assign_attrs(description="column integrated water vapor")
    )
    physics_precip = (
        state[PRECIP_RATE]
        .assign_attrs(units="kg/m^2/s")
        .assign_attrs(
            description="surface precipitation rate due to parameterized physics"
        )
    )

    diags.update(
        {
            "column_moistening_nudge": net_moistening,
            "column_heating_nudge": net_heating,
            "water_vapor_path": water_vapor_path,
            "physics_precip": physics_precip,
        }
    )

    if ("x_wind" in nudging_tendency.keys()) and ("y_wind" in nudging_tendency.keys()):
        wind_rotation_matrix = catalog["wind_rotation/c48"].to_dask()
        u_tendency, v_tendency = center_and_rotate_xy_winds(
            wind_rotation_matrix, nudging_tendency["x_wind"], nudging_tendency["y_wind"]
        )
        rotation_mapping = {
            "eastward_wind": u_tendency,
            "northward_wind": v_tendency,
        }
        for a_name, tendency in rotation_mapping.items():
            integrated_wind_tendency = _mass_average(tendency, state[DELP], "z")
            diags[
                f"column_{a_name}_tendency_nudge"
            ] = integrated_wind_tendency.assign_attrs(
                units="m s^-2",
                description=(
                    f"column mass-averaged {a_name} wind tendency due to nudging"
                ),
            )

    if DELP in nudging_tendency:
        net_mass_tendency = (
            (nudging_tendency[DELP] / gravity)
            .sum("z")
            .assign_attrs(
                units="kg/m^2/s",
                description="column-integrated mass tendency due to nudging",
            )
        )
        diags["column_mass_tendency_nudge"] = net_mass_tendency
    diags.update(_append_key_label(nudging_tendency, label))

    return diags


def _append_key_label(d: Diagnostics, suffix: str) -> Diagnostics:
    return_dict: Diagnostics = {}
    for key, value in d.items():
        return_dict[str(key) + suffix] = value
    return return_dict


def _mass_average(
    da: xr.DataArray, delp: xr.DataArray, vertical_dim: str = "z"
) -> xr.DataArray:
    total_thickness = delp.sum(vertical_dim)
    mass_average = (da * delp).sum(vertical_dim) / total_thickness
    mass_average = mass_average.assign_attrs(**da.attrs)
    return mass_average
