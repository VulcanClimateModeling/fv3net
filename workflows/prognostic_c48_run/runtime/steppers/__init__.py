from .reservoir import HybridReservoirConfig, get_reservoir_steppers
