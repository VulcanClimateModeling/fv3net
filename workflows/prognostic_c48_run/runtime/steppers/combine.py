from collections import Counter
from typing import Optional, List, Union, MutableMapping, Hashable
import xarray as xr

from runtime.steppers.machine_learning import PureMLStepper
from runtime.steppers.prescriber import Prescriber


def _merge_outputs(
    outputs: List[MutableMapping[Hashable, xr.DataArray]]
) -> MutableMapping[Hashable, xr.DataArray]:
    return {k: v for d in outputs for k, v in d.items()}


class CombinedStepper:
    def __init__(self, steppers: Optional[List[Union[Prescriber, PureMLStepper]]]):
        self._steppers = steppers
        # have to call the steppers first before checking if they collide
        self._verified_no_collisions = False

    def __call__(self, time, state):
        tendencies, diagnostics, state_updates = [], [], []
        for stepper in self._steppers:
            t, d, s = stepper(time, state)
            tendencies.append(t)
            diagnostics.append(d)
            state_updates.append(s)
        if self._verified_no_collisions is False:
            for outputs in [tendencies, diagnostics, state_updates]:
                self._check_for_collisions(outputs)
        return (
            _merge_outputs(tendencies),
            _merge_outputs(diagnostics),
            _merge_outputs(state_updates),
        )

    def _check_for_collisions(
        self, outputs: List[MutableMapping[Hashable, xr.DataArray]]
    ):
        output_keys = [list(output) for output in outputs]
        all_keys: List = sum(output_keys, [])
        key_counts = Counter(all_keys)
        collisions = []
        for key, count in key_counts.items():
            if count > 1:
                collisions.append(key)
        if len(collisions) > 0:
            raise ValueError(
                "Steppers configured in prephysics state updates have overlapping "
                f"update keys: {collisions}"
            )
        else:
            self._verified_no_collisions = True
