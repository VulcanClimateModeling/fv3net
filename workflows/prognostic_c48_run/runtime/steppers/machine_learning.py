"""Code for machine Learning in prognostic runs
"""
import dataclasses
import logging
import os
from typing import Hashable, Iterable, Mapping, Sequence, Set, Tuple, cast

import fv3fit
import xarray as xr
from runtime.diagnostics import compute_diagnostics, compute_ml_momentum_diagnostics
from runtime.names import DELP, SPHUM
from runtime.types import Diagnostics, State
from vcm import thermo
import vcm

__all__ = ["MachineLearningConfig", "PureMLStepper", "open_model"]


logger = logging.getLogger(__name__)

NameDict = Mapping[Hashable, Hashable]


@dataclasses.dataclass
class MachineLearningConfig:
    """Machine learning configurations

    Attributes:
        model: list of URLs to fv3fit models.
        diagnostic_ml: do not apply ML tendencies if true.
        input_standard_names: mapping from non-standard names to the standard
            ones used by the model. Renames the ML inputs. Useful if the
            input variables in the ML model are inconsistent with
            the canonical names used in the wrapper.
        output_standard_names: mapping from non-standard names to the standard
            ones used by the model. Renames the ML predictions.

    Example::

        MachineLearningConfig(
            model=["gs://vcm-ml-scratch/test-annak/ml-pipeline-output"],
            diagnostic_ml=False,
            input_standard_names={},
            output_standard_names={},
        )

    """

    model: Sequence[str] = dataclasses.field(default_factory=list)
    diagnostic_ml: bool = False
    input_standard_names: Mapping[Hashable, Hashable] = dataclasses.field(
        default_factory=dict
    )
    output_standard_names: Mapping[Hashable, Hashable] = dataclasses.field(
        default_factory=dict
    )


def non_negative_sphum(
    sphum: xr.DataArray, dQ1: xr.DataArray, dQ2: xr.DataArray, dt: float
) -> Tuple[xr.DataArray, xr.DataArray]:
    delta = dQ2 * dt
    reduction_ratio = (-sphum) / (dt * dQ2)  # type: ignore
    dQ1_updated = xr.where(sphum + delta >= 0, dQ1, reduction_ratio * dQ1)
    dQ2_updated = xr.where(sphum + delta >= 0, dQ2, reduction_ratio * dQ2)
    return dQ1_updated, dQ2_updated


def _invert_dict(d: Mapping) -> Mapping:
    return dict(zip(d.values(), d.keys()))


class RenamingAdapter:
    """Adapter object for renaming model variables

    Attributes:
        model: a model to rename
        rename_in: mapping from standard names to input names of model
        rename_out: mapping from standard names to the output names of model

    """

    def __init__(
        self, model: fv3fit.Predictor, rename_in: NameDict, rename_out: NameDict = None
    ):
        self.model = model
        self.rename_in = rename_in
        self.rename_out = {} if rename_out is None else rename_out

    def _rename(self, ds: xr.Dataset, rename: NameDict) -> xr.Dataset:

        all_names = set(ds.dims) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        redimed = ds.rename_dims(rename_restricted)

        all_names = set(ds.data_vars) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        return redimed.rename(rename_restricted)

    def _rename_inputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, self.rename_in)

    def _rename_outputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, _invert_dict(self.rename_out))

    @property
    def input_variables(self) -> Set[str]:
        invert_rename_in = _invert_dict(self.rename_in)
        return {invert_rename_in.get(var, var) for var in self.model.input_variables}

    def predict_columnwise(self, arg: xr.Dataset, **kwargs) -> xr.Dataset:
        input_ = self._rename_inputs(arg)
        prediction = self.model.predict_columnwise(input_, **kwargs)
        return self._rename_outputs(prediction)


class MultiModelAdapter:
    def __init__(self, models: Iterable[RenamingAdapter]):
        self.models = models

    @property
    def input_variables(self) -> Set[str]:
        vars = [model.input_variables for model in self.models]
        return {var for model_vars in vars for var in model_vars}

    def predict_columnwise(self, arg: xr.Dataset, **kwargs) -> xr.Dataset:
        predictions = []
        for model in self.models:
            predictions.append(model.predict_columnwise(arg, **kwargs))
        return xr.merge(predictions)


def open_model(config: MachineLearningConfig) -> MultiModelAdapter:
    model_paths = config.model
    models = []
    for path in model_paths:
        model = fv3fit.load(path)
        rename_in = config.input_standard_names
        rename_out = config.output_standard_names
        models.append(RenamingAdapter(model, rename_in, rename_out))
    return MultiModelAdapter(models)


def download_model(config: MachineLearningConfig, path: str) -> Sequence[str]:
    """Download models to local path and return the local paths"""
    remote_model_paths = config.model
    local_model_paths = []
    for i, remote_path in enumerate(remote_model_paths):
        local_path = os.path.join(path, str(i))
        os.makedirs(local_path)
        fs = vcm.cloud.get_fs(remote_path)
        fs.get(remote_path, local_path, recursive=True)
        local_model_paths.append(local_path)
    return local_model_paths


def predict(model: MultiModelAdapter, state: State) -> State:
    """Given ML model and state, return prediction"""
    state_loaded = {key: state[key] for key in model.input_variables}
    ds = xr.Dataset(state_loaded)  # type: ignore
    output = model.predict_columnwise(ds, feature_dim="z")
    return {key: cast(xr.DataArray, output[key]) for key in output.data_vars}


class PureMLStepper:

    label = "machine_learning"

    def __init__(self, model: MultiModelAdapter, timestep: float):
        self.model = model
        self.timestep = timestep

    def __call__(self, time, state):

        diagnostics: Diagnostics = {}
        delp = state[DELP]

        tendency: State = predict(self.model, state)

        dQ1_initial = tendency.get("dQ1", xr.zeros_like(state[SPHUM]))
        dQ2_initial = tendency.get("dQ2", xr.zeros_like(state[SPHUM]))

        dQ1_updated, dQ2_updated = non_negative_sphum(
            state[SPHUM], dQ1_initial, dQ2_initial, dt=self.timestep,
        )

        if "dQ1" in tendency:
            diag = thermo.column_integrated_heating(dQ1_updated - tendency["dQ1"], delp)
            diagnostics.update(
                {"column_integrated_dQ1_change_non_neg_sphum_constraint": (diag)}
            )
            tendency.update({"dQ1": dQ1_updated})
        if "dQ2" in tendency:
            diag = thermo.mass_integrate(dQ2_updated - tendency["dQ2"], delp, dim="z")
            diag = diag.assign_attrs({"units": "kg/m^2/s"})
            diagnostics.update(
                {"column_integrated_dQ2_change_non_neg_sphum_constraint": (diag)}
            )
            tendency.update({"dQ2": dQ2_updated})

        diagnostics["specific_humidity_limiter_active"] = xr.where(
            dQ2_initial != dQ2_updated, 1, 0
        )

        state_updates = {}
        return (
            tendency,
            diagnostics,
            state_updates,
        )

    def get_diagnostics(self, state, tendency):
        diags = compute_diagnostics(state, tendency, self.label)
        return diags, diags[f"net_moistening_due_to_{self.label}"]

    def get_momentum_diagnostics(self, state, tendency):
        return compute_ml_momentum_diagnostics(state, tendency)


class MLStateStepper(PureMLStepper):
    def __call__(self, time, state):

        diagnostics: Diagnostics = {}
        state_updates: State = predict(self.model, state)

        for name in state_updates.keys():
            diagnostics[name] = state_updates[name]

        tendency = {}
        return (
            tendency,
            diagnostics,
            state_updates,
        )
