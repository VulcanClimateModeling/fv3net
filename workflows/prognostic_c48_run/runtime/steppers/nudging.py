import functools

import fv3gfs.util
import fv3gfs.wrapper
from runtime.diagnostics.machine_learning import (
    compute_baseline_diagnostics,
    compute_nudging_diagnostics,
)
from runtime.nudging import (
    NudgingConfig,
    get_nudging_tendency,
    get_reference_surface_temperatures,
    nudging_timescales_from_dict,
    setup_get_reference_state,
)
from runtime.types import Diagnostics, State
from runtime.names import SST, TSFC


class PureNudger:

    net_moistening = "net_moistening_due_to_nudging"

    def __init__(
        self, config: NudgingConfig, communicator: fv3gfs.util.CubedSphereCommunicator,
    ):

        variables_to_nudge = list(config.timescale_hours)
        self._get_reference_state = setup_get_reference_state(
            config,
            variables_to_nudge + [SST, TSFC],
            fv3gfs.wrapper.get_tracer_metadata(),
            communicator,
        )

        self._nudging_timescales = nudging_timescales_from_dict(config.timescale_hours)
        self._get_nudging_tendency = functools.partial(
            get_nudging_tendency, nudging_timescales=self._nudging_timescales,
        )

    def __call__(self, time, state):
        reference = self._get_reference_state(time)
        tendencies = get_nudging_tendency(state, reference, self._nudging_timescales)
        ssts = get_reference_surface_temperatures(state, reference)

        reference = {
            f"{key}_reference": reference_state
            for key, reference_state in reference.items()
        }
        return tendencies, reference, ssts

    def get_diagnostics(self, state: State, tendency: State) -> Diagnostics:
        diags = {}
        diags.update(compute_baseline_diagnostics(state))
        diags.update(compute_nudging_diagnostics(state, tendency))
        return diags

    def get_momentum_diagnostics(self, state: State, tendency: State) -> Diagnostics:
        return {}
