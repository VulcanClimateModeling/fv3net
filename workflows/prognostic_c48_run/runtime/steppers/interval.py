import cftime
import dataclasses
from datetime import timedelta
from typing import Tuple, Union, Optional, List
import xarray as xr
import logging
import vcm

from runtime.types import Diagnostics
from runtime.steppers.stepper import Stepper
from runtime.steppers.machine_learning import MachineLearningConfig
from runtime.steppers.prescriber import PrescriberConfig
from runtime.nudging import NudgingConfig
from runtime.diagnostics.compute import precipitation_sum, precipitation_rate
from runtime.names import SPHUM, DELP, TOTAL_PRECIP, TOTAL_PRECIP_RATE

logger = logging.getLogger(__name__)


@dataclasses.dataclass
class IntervalConfig:
    """Configuration for interval steppers

    base_config: config for the stepper used at the end of the interval
    apply_interval_seconds: interval in seconds
    offset_seconds: offset from the start of the run in seconds to count
        as start of intervals
    record_fields_before_update: fields listed here will have their values
        immediately before the stepper update recorded in diagnostics
    n_calls: if provided, stop after this many calls to the stepper
        useful for synchronizing reservoir models at the start of runs
    """

    base_config: Union[PrescriberConfig, MachineLearningConfig, NudgingConfig]
    apply_interval_seconds: int
    offset_seconds: int = 0
    record_fields_before_update: Optional[List[str]] = None
    n_calls: Optional[int] = None
    do_total_precip_update: bool = True


class IntervalStepper:
    def __init__(
        self,
        apply_interval_seconds: float,
        stepper: Stepper,
        offset_seconds: float = 0,
        n_calls: Optional[int] = None,
        record_fields_before_update: Optional[List[str]] = None,
        do_total_precip_update: bool = True,
    ):
        self.start_time = None
        self.interval = timedelta(seconds=apply_interval_seconds)
        self.stepper = stepper
        self.offset_seconds = timedelta(seconds=offset_seconds)
        self._record_fields_before_update = record_fields_before_update or []
        self.n_calls = n_calls
        self._call_count = 0
        self._do_total_precip_update = do_total_precip_update

    @property
    def label(self):
        """Label used for naming diagnostics.
        """
        return f"interval_{self.stepper.label}"

    def _need_to_update(self, time: cftime.DatetimeJulian):
        if self.start_time is not None:
            if (
                (time - self.start_time - self.offset_seconds) % self.interval
            ).seconds != 0:
                return False
            else:
                if self.n_calls is None:
                    return True
                else:
                    if self._call_count < self.n_calls:
                        return True
                    else:
                        return False

        else:
            logger.info(f"Setting interval stepper start time to {time}")
            self.start_time = time
            return False

    def get_diagnostics_prior_to_update(self, state):
        return {
            f"{key}_before_interval_update": state[key]
            for key in self._record_fields_before_update
        }

    def __call__(self, time, state):

        if self._need_to_update(time) is False:
            # Diagnostic must be available at all timesteps, not just when
            # the base stepper is called
            diags = self.get_diagnostics_prior_to_update(state)
            diags.update(
                {
                    f"net_moistening_due_to_{self.label}": xr.zeros_like(
                        state[TOTAL_PRECIP]
                    )
                }
            )
            return {}, diags, {}
        else:
            logger.info(f"applying interval stepper at time {time}")
            tendencies, diagnostics, state_updates = self.stepper(time, state)
            diagnostics.update(self.get_diagnostics_prior_to_update(state))
            self._call_count += 1

            if self._do_total_precip_update and SPHUM in state_updates:
                logger.info(f"updating total precip at time {time}")
                net_moistening = self._net_moistening(
                    sphum_before_update=state[SPHUM],
                    sphum_after_update=state_updates[SPHUM],
                    delp=state[DELP],
                )
                diagnostics[f"net_moistening_due_to_{self.label}"] = net_moistening
                total_precipitation_update = precipitation_sum(
                    state[TOTAL_PRECIP], net_moistening, self.interval.total_seconds()
                )
                diagnostics[TOTAL_PRECIP_RATE] = precipitation_rate(
                    total_precipitation_update, self.interval.total_seconds()
                )
                state_updates[TOTAL_PRECIP] = total_precipitation_update
            else:
                logger.info(f"Not updating total precip  ")
            return tendencies, diagnostics, state_updates

    def _net_moistening(self, sphum_before_update, sphum_after_update, delp):
        _dQ2 = (
            sphum_after_update - sphum_before_update
        ) / self.interval.total_seconds()
        return vcm.mass_integrate(_dQ2, delp, "z")

    def get_diagnostics(self, state, tendency) -> Tuple[Diagnostics, xr.DataArray]:
        diags, moistening = self.stepper.get_diagnostics(state, tendency)
        diags.update(self.get_diagnostics_prior_to_update(state))
        return diags, moistening
