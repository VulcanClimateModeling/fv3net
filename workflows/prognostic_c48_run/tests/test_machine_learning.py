from runtime.steppers.machine_learning import PureMLStepper
from machine_learning_mocks import get_mock_predictor
import requests
import xarray as xr
import yaml
import pytest
import vcm.testing


@pytest.fixture(scope="session")
def state(tmp_path_factory):
    url = "https://github.com/VulcanClimateModeling/vcm-ml-example-data/blob/b100177accfcdebff2546a396d2811e32c01c429/fv3net/prognostic_run/inputs_4x4.nc?raw=true"  # noqa
    r = requests.get(url)
    lpath = tmp_path_factory.getbasetemp() / "input_data.nc"
    lpath.write_bytes(r.content)
    return xr.open_dataset(str(lpath))


@pytest.fixture(params=["tendencies", "rad_fluxes"])
def model_output_type(request):
    return request.param


@pytest.fixture
def ml_stepper(model_output_type):
    timestep = 900
    mock_model = get_mock_predictor(model_output_type)
    ml_stepper = PureMLStepper(mock_model, timestep, hydrostatic=False)
    return ml_stepper


@pytest.mark.parametrize("output_type", ["tendencies", "rad_fluxes"])
def test_mock_predictor_checksum(output_type, state, regtest):
    mock_model = get_mock_predictor(output_type)
    predicted = mock_model.predict(state)
    test_state_regression(predicted, regtest)


def test_ml_steppers_schema_unchanged(state, ml_stepper, regtest):
    (tendencies, diagnostics, states) = ml_stepper(None, state)
    xr.Dataset(diagnostics).info(regtest)
    xr.Dataset(tendencies).info(regtest)
    xr.Dataset(states).info(regtest)


def test_state_regression(state, regtest):
    checksum = vcm.testing.checksum_dataarray_mapping(state)
    print(checksum, file=regtest)


def test_ml_steppers_regression_checksum(state, ml_stepper, regtest):
    (tendencies, diagnostics, states) = ml_stepper(None, state)
    checksums = yaml.safe_dump(
        [
            ("tendencies", vcm.testing.checksum_dataarray_mapping(tendencies)),
            ("diagnostics", vcm.testing.checksum_dataarray_mapping(diagnostics)),
            ("states", vcm.testing.checksum_dataarray_mapping(states)),
        ]
    )

    print(checksums, file=regtest)


@pytest.mark.parametrize(
    "mock_predictor", ["state_and_tendency", "state_tendency_and_intermediate"]
)
def test_ml_stepper_state_update(state, mock_predictor):
    ml_stepper = PureMLStepper(
        get_mock_predictor(mock_predictor), timestep=900, hydrostatic=False
    )
    (tendencies, diagnostics, states) = ml_stepper(None, state)
    assert set(states) == {"total_sky_downward_shortwave_flux_at_surface"}
    assert set(tendencies) == {"dQ1"}
    if mock_predictor == "state_tendendency_and_intermediate":
        assert "shortwave_transmissivity_of_atmospheric_column" in diagnostics
