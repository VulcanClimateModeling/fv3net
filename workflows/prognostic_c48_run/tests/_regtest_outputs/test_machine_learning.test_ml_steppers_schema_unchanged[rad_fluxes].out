xarray.Dataset {
dimensions:
	y = 4 ;
	x = 4 ;
	z = 63 ;

variables:
	float64 total_sky_downward_shortwave_flux_at_surface(y, x) ;
	float64 total_sky_downward_longwave_flux_at_surface(y, x) ;
	int64 specific_humidity_limiter_active(z, y, x) ;

// global attributes:
}xarray.Dataset {
dimensions:

variables:

// global attributes:
}xarray.Dataset {
dimensions:
	y = 4 ;
	x = 4 ;

variables:
	float64 total_sky_downward_shortwave_flux_at_surface(y, x) ;
	float64 total_sky_downward_longwave_flux_at_surface(y, x) ;

// global attributes:
}